### PR TITLE
[ExportVerilog] Emit constants as localparam

### LINKED ISSUE #884: Warning that Verilog may cause mismatches between simulation and synthesis
The formal verification tool flags this as a warning (not error).  I have attached the MLIR output and a corrected version below.

Warning: Variable(s) is(are) being read asynchronously. This may cause simulation-synthesis mismatches. (Signal: _T_0 Block: testcase_mfc.sv Line: 43) 
The problem with this description is that the simulation model does not react to changes in reset_data while reset_enable is held high.
In Verilog, you can't add _T_0 to the sensitivity list without breaking the model. You must verify that your model does not rely on the fact that during reset, the simulator holds tmp50 steady even though _TMP_50 changes. The synthesized register will be transparent during reset.

Or change the RTL, to have constant driving reset line.

_The FIR file_circuit top_mod :
```verilog
  module top_mod :
    input clock: Clock
    input reset: UInt<1>
    input arst: AsyncReset
    input inp_i: SInt<170>
    output out_i: SInt<170>
    reg tmp50: SInt<2>, clock with: (reset => (arst, SInt<2>(0)))
    tmp50 <= inp_i
    out_i <= tmp50
```

**The output that produces the warning (just the relevant section):
```verilog
 always_ff @(posedge clock or posedge arst**) begin      // testcase.fir:9:5
    if (arst)   // testcase.fir:9:5
      **tmp50 <= _T_0;    // testcase.fir:9:5**
    else        // testcase.fir:9:5
      tmp50 <= _T;      // testcase.fir:10:11
  end // always_ff @(posedge or posedge)
```

**The structure that is read without warnings:**
```verilog
module top_mod(
  input          clock,
  input          reset,
  input          arst,
  input  [169:0] inp_i,
  output [169:0] out_i
);
`ifdef RANDOMIZE_REG_INIT
  reg [31:0] _RAND_0;
`endif // RANDOMIZE_REG_INIT
  reg [1:0] tmp50;
  assign out_i = {{168{tmp50[1]}},tmp50};
  always @(posedge clock or posedge arst) begin
    if (arst) begin
      **tmp50 <= 2'sh0;**
    end else begin
      tmp50 <= inp_i[1:0];
    end
  end
```

### PR DESCRIPTION
**Pull request by fabianschuiki**
Emit `rtl.constant` operations as `localparam` when we decide to print them as out-of-line expressions. This requires additional special treatment since the localparam cannot be reassigned and the value must be provided already at the declaration site.

When `rtl.constant` operations are used before defined, move the definition up to the beginning of the block, similar to wire/reg. This avoids introducing a temporary dynamic wire and emits the constant as a localparam.

This fixes #884 by moving the constant reset values of always_ff statements into localparams.

### CODE REVIEW

**Review by darthscsi — COMMENTED** [review]
I think we want something a little more general for constants (and constantX and constantZ).  They generally should be inline no matter their users, unless they are really long, in which case we need to spill them to a localparam.

**Comment by fabianschuiki** [comment]
@darthscsi Reworked this a bit to make it more general: Constants are now always spilled into `localparam`s. This requires a few tweaks to the emission of declarations for out-of-line temporaries, since the value needs to be provided immediately with the declaration. I also disabled the decision to not inline across blocks for constants, as they do not have any other operands. That now allows them to almost always be inlined; unless they're explicitly outlined through the usual heuristic, in which case they spill into a localparam.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 1]
```diff
@@ -1319,8 +1326,9 @@ static bool isExpressionUnableToInline(Operation *op) {
   for (auto user : op->getUsers()) {
     // If the user is in a different block and the op shouldn't be inlined, then
     // we emit this as an out-of-line declaration into its block and the user
-    // can refer to it.
-    if (user->getBlock() != opBlock)
+    // can refer to it. Does not apply to constants, which can always be inlined
+    // into other blocks.
+    if (user->getBlock() != opBlock && !isConstantExpression(op))
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1331]: Is this "required" for correctness or are you just trying to clean up the generated IR?

If this is for prettiness, I think we should handle this in rtlcleanup and shouldn't be part of this patch.  If we need it for some reason, I'd love to know why.  Also, in that case, please hoist `isConstantExpresssion(op)` out of the getUsers loop since it is loop invariant

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 2]
```diff
@@ -1623,6 +1631,11 @@ void StmtEmitter::emitExpression(Value exp,
 
   // Emit each stmt expression in turn.
   for (auto *expr : tooLargeSubExpressions) {
+    // Do not emit assignments for constants, as these will be printed
+    // immediately after declaration of their temporary.
+    if (isConstantExpression(expr) &&
+        tooLargeSubExpressions[0]->getParentOp()->hasTrait<ProceduralRegion>())
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1637]: Please remove this change too, it shouldn't be needed if I understand correctly.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 3]
```diff
@@ -1640,8 +1653,18 @@ void StmtEmitter::emitExpression(Value exp,
                     outBuffer.end());
     outBuffer.resize(blockDeclarationInsertPointIndex);
     for (auto *expr : tooLargeSubExpressions) {
+      SmallPtrSet<Operation *, 8> emittedExprs;
       emitDeclarationForTemporary(expr);
-      os << ";\n";
+
+      // If this is a constant, we have to immediately assign its value since
+      // as required per the `localparam` construct they are emitted as.
+      if (isConstantExpression(expr)) {
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1661]: Should this special case go in `emitDeclarationForTemporary`?

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -2354,6 +2386,12 @@ void StmtEmitter::collectNamesEmitDecls(Block &block) {
       printUnpackedTypePostfix(type, os);
     }
 
+    // Constants carry their assignment directly in the declaration.
+    if (isConstantExpression(decl)) {
+      os << " = ";
+      emitExpression(decl->getResult(0), ops, ForceEmitMultiUse);
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2392]: Does merging this into emitDeclarationForTemporary eliminate the need for this?

File: test/ExportVerilog/rtl-dialect.mlir [hunk 5]
```diff
@@ -599,11 +599,10 @@ rtl.module @longvariadic(%a: i8) -> (%b: i8) {
 // CHECK-EMPTY:
 // CHECK-NEXT:   reg memory_r_en_pipe[0:0];
 // CHECK-EMPTY:
-// CHECK-NEXT:   wire _T = 1'h0;
 // CHECK-NEXT:   always_ff @(posedge clock)
-// CHECK-NEXT:     memory_r_en_pipe[_T] <= _T;
+// CHECK-NEXT:     memory_r_en_pipe[1'h0] <= 1'h0;
 // CHECK-NEXT:   initial
-// CHECK-NEXT:     memory_r_en_pipe[_T] = _T;
+// CHECK-NEXT:     memory_r_en_pipe[1'h0] = 1'h0;
```
Comment (lattner) [inline test/ExportVerilog/rtl-dialect.mlir:605]: This change shouldn't happen, we should emit the 1 out of line to match the IR.  A future unrelated patch to RTLCleanup should handle this, since it is just a prettification

**Review by lattner — COMMENTED** [review]
Thank you for working on this.  Please try to keep the verilog emitter as simple as possible, while always producing correct output.

File: test/ExportVerilog/rtl-dialect.mlir (hunk 5 above)
Comment (fabianschuiki) [reply test/ExportVerilog/rtl-dialect.mlir:605]: Sure thing. Also the rationale for not inlining discussed in #895 makes sense. Will remove this.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:1331]: Noep this isn't needed for correctness. It was merely here to encourage inlining of constants. But with the rationale that we should rather push duplicates of constants into the blocks where they're used if we want them to be inlined, this change isn't needed :+1: Will drop.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 3 above)
Comment (fabianschuiki) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:1661]: Yes this simplifies things. Also removes the need for the special case you commented on above.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 4 above)
Comment (fabianschuiki) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2392]: Not directly since this part of the emitter (`StmtEmitter::collectNamesEmitDecls`) does not use `emitDeclarationForTemporary`. We would have to extend `emitDeclarationForTemporary` to also cover the use case here (zero bit type handling, type string padding, interface instantiation, unpacked type postfixes). Not sure if that would make things clearer -- reduces some duplication though.

**Comment by fabianschuiki** [comment]
Applied @lattner's suggestions. Also added support for moving constants up in procedural blocks if they are used before defined, like wire/reg which gets rid of unnecessary non-constant temporaries.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 6]
```diff
@@ -1507,14 +1514,29 @@ class StmtEmitter : public EmitterBase,
...
+  bool emitDeclarationForTemporary(Operation *op) {
     indent() << getVerilogDeclWord(op) << " ";
     if (printPackedType(stripUnpackedTypes(op->getResult(0).getType()), os,
                         op->getLoc()))
       os << ' ';
     os << emitter.getName(op->getResult(0));
+
+    // If this is a constant, we have to immediately assign its value as
+    // required by the `localparam` construct.
+    if (isConstantExpression(op)) {
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:1530]: Plz invert the condition and use early out to reduce the nesting.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp (hunk 4 above)
Comment (lattner) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:2389]: ok, lets leave it as-is for now, thx

**Review by lattner — APPROVED** [review]
Looks good, thanks!

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 7]
```diff
@@ -56,6 +56,11 @@ static bool isExpressionAlwaysInline(Operation *op) {
   return false;
 }
 
+/// Return whether an operation is a constant.
+static bool isConstantExpression(Operation *op) {
+  return isa<ConstantOp>(op) || isa<ConstantXOp>(op) || isa<ConstantZOp>(op);
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Translation/ExportVerilog/ExportVerilog.cpp:61]: Yes absolutely. Any of the constructs where you can emit the type (`wire`, `reg`, `var`, `localparam`, `parameter`, `input`, `output`, `inout`) will implicitly default to the 4-state `logic` type in most cases (tons of special cases aside). So X/Z is totally fine here.

**Comment by lattner** [comment]
nice, thanks, lgtm!
